### PR TITLE
deleteCallback to delete different versions of a file

### PR DESCRIPTION
**Pull request by jorisvaesen**
Ref #453 

I'll add an example to the docs if this feature gets approved.

### CODE REVIEW

File: docs/configuration.rst [hunk 1]
```diff
@@ -76,6 +76,18 @@ passed in under each field in your behavior configuration.
 
    -  Default: (boolean) ``true``
 
+-  ``deleteCallback``: A callable that can be used to delete different versions of the file.
+
+   -  Default: (boolean) ``false``
+   -  Available arguments:
+
+      -  ``string $path``: Basepath of the file you want to delete
+      -  ``object $entity``: The entity you want to delete
```
Comment (davidyell) [inline docs/configuration.rst]: I would actually specify 'Entity' instead of 'object' here, just to be clearer.

File: tests/TestCase/Model/Behavior/UploadBehaviorTest.php [hunk 2]
```diff
@@ -28,7 +28,8 @@ public function setup()
                 'error' => UPLOAD_ERR_OK,
                 'size' => 1,
                 'type' => 'text',
-                'keepFilesOnDelete' => false
+                'keepFilesOnDelete' => false,
+                'deleteCallback' => false
```
Comment (davidyell) [inline tests/TestCase/Model/Behavior/UploadBehaviorTest.php]: My personal preference would be `null` here to demonstrate a 'nothing' value, rather than a boolean.

File: tests/TestCase/Model/Behavior/UploadBehaviorTest.php [hunk 3]
```diff
@@ -442,6 +443,50 @@ public function testAfterDeleteUsesPathProcessorToDetectPathToTheFile()
...
+            return [
+                $path . $entity->{$field},
+                $path . 'sm-' . $entity->{$field},
+                $path . 'lg-' . $entity->{$field}
+            ];
+        };
+
+        $behavior->config($this->dataOk);
+
+        // expecting getPathProcessor to be called with right arguments for dataOk
```
Comment (davidyell) [inline tests/TestCase/Model/Behavior/UploadBehaviorTest.php]: Not sure about committing comments to the repo, but the repo code style will be up to @josegonzalez

**Comment by davidyell** [comment]
I would also probably throw in a test case for when `deleteCallback` is `false` or `null` just to ensure it's not executed, which might help prevent future regressions.

Otherwise, reading the code, it seems okay to me.

File: docs/configuration.rst (hunk 1 above)
Comment (jorisvaesen) [inline docs/configuration.rst]: Indeed.

File: tests/TestCase/Model/Behavior/UploadBehaviorTest.php (hunk 2 above)
Comment (jorisvaesen) [inline tests/TestCase/Model/Behavior/UploadBehaviorTest.php]: Just saw nameCallback also uses null as default, will change this accordingly.

File: tests/TestCase/Model/Behavior/UploadBehaviorTest.php (hunk 3 above)
Comment (jorisvaesen) [inline tests/TestCase/Model/Behavior/UploadBehaviorTest.php]: Hmm, it's just because I copied code from https://github.com/FriendsOfCake/cakephp-upload/blob/master/tests/TestCase/Model/Behavior/UploadBehaviorTest.php#L415
Will delete the comments.

Comment (davidyell) [inline tests/TestCase/Model/Behavior/UploadBehaviorTest.php]: Might not be a big deal as it's in a test. Although I do wonder if you have to comment each assert, does that mean they could be different test cases? Just a thought.

**Comment by josegonzalez** [comment]
Seems fine, needs an example.

**Comment by saeideng** [comment]
nice return
```
                     return [
                            $path . $entity->{$field},
                           $path . 'thumbnail-' . $entity->{$field}
                       ];
```
good job :+1:
